homeassistant/helpers/entity_component.py

```diff
@@ -50,40 +53,75 @@ def setup(self, config):
         Loads the platforms from the config and will listen for supported
         discovered platforms.
         """
+        run_coroutine_threadsafe(
+            self.async_setup(config), self.hass.loop
+        ).result()
+
+    @asyncio.coroutine
+    def async_setup(self, config):
+        """Set up a full entity component.
+
+        Loads the platforms from the config and will listen for supported
+        discovered platforms.
+
+        This method must be run in the event loop.
+        """
         self.config = config
 
         # Look in config for Domain, Domain 2, Domain 3 etc and load them
+        tasks = []
         for p_type, p_config in config_per_platform(config, self.domain):
-            self._setup_platform(p_type, p_config)
+            tasks.append(self.hass.loop.create_task(
+                self._setup_platform(p_type, p_config)
+            ))
+        yield from asyncio.gather(*tasks, loop=self.hass.loop)
 
         # Generic discovery listener for loading platform dynamically
         # Refer to: homeassistant.components.discovery.load_platform()
+        @callback
         def component_platform_discovered(platform, info):
             """Callback to load a platform."""
-            self._setup_platform(platform, {}, info)
+            self.hass.async_run_job(self._setup_platform, platform, {}, info)
```

Comment:
You should be able to call `hass.loop.create_task` directly because you know it will always be a coroutine.